### PR TITLE
Fix large/small scroll in ScrollViewer.

### LINKED ISSUE #3245: ScrollViewer doesnt behave as expected when clicking the scroll buttons or the track
Test Code

```
<ScrollViewer HorizontalScrollBarVisibility="Visible">
            <StackPanel Height="10000" Width="10000">
                <TextBlock Text="test" />
                <TextBlock Text="test" />
                <TextBlock Text="test" />

                <TextBlock Text="test" />
                <TextBlock Text="test" />
                <TextBlock Text="test" />
                <TextBlock Text="test" />
            </StackPanel>
        </ScrollViewer>
```
wpf behaves like:
wpf-scrollviewer

Avalonia like:
avalonia-scrollviewer

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Fixes scrolling in a `ScrollViewer` when clicking on the scrollbar buttons or on the track.

## What is the current behavior?

The `ScrollViewer` always scrolls by one pixel when clicking the scrollbar buttons or the track.

## What is the updated/expected behavior with this PR?

For non-logical scrolling:

- Use 16 for small scroll size (value taken from WPF)
- Use viewport size for large scroll

For logical scrolling, use the `ScrollSize`/`PageScrollSize` defined on `ILogicalScrollable`. Note that this required a small breaking change to `ILogicalScrollable`.

## Checklist

- [x] Added unit tests (if possible)?

## Breaking changes

Required a small breaking change to `ILogicalScrollable`: 

- `InvalidateScroll` property has become an event: `ScrollInvalidated`
- A `RaiseScrollInvalidated` method has been added to raise `ScrollInvalidated`

## Fixed issues

Fixes #3245

### CODE REVIEW

**Comment by danwalmsley** [comment]
@grokys is this a backport candidate?

File: Avalonia.sln [hunk 1]
```diff
@@ -204,16 +204,16 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Avalonia.Dialogs", "src\Ava
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Avalonia.FreeDesktop", "src\Avalonia.FreeDesktop\Avalonia.FreeDesktop.csproj", "{4D36CEC8-53F2-40A5-9A37-79AAE356E2DA}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Avalonia.Controls.DataGrid.UnitTests", "tests\Avalonia.Controls.DataGrid.UnitTests\Avalonia.Controls.DataGrid.UnitTests.csproj", "{351337F5-D66F-461B-A957-4EF60BDB4BA6}"
```
Comment (grokys) [inline Avalonia.sln:207]: VS 16.5 keeps insisting on this change. We need to push it at some point, so might as well push it here.

File: src/Avalonia.Controls/Presenters/ItemVirtualizer.cs [hunk 2]
```diff
@@ -99,9 +100,14 @@ public Size Extent
         {
             get
             {
-                return Vertical ?
-                    new Size(Owner.Panel.DesiredSize.Width, ExtentValue) :
-                    new Size(ExtentValue, Owner.Panel.DesiredSize.Height);
+                if (IsLogicalScrollEnabled)
```
Comment (grokys) [inline src/Avalonia.Controls/Presenters/ItemVirtualizer.cs:103]: The checks here for `IsLogicalScrollEnabled` aren't strictly needed, but they stop exceptions being thrown when viewing a non-virtualized `ItemsPresenter` in DevTools.

File: src/Avalonia.Controls/Presenters/ItemVirtualizer.cs [hunk 3]
```diff
@@ -164,10 +185,10 @@ public static ItemVirtualizer Create(ItemsPresenter owner)
             }
 
             var virtualizingPanel = owner.Panel as IVirtualizingPanel;
-            var scrollable = (ILogicalScrollable)owner;
+            var scrollContentPresenter = owner.Parent as IScrollable;
             ItemVirtualizer result = null;
 
-            if (virtualizingPanel != null && scrollable.InvalidateScroll != null)
```
Comment (grokys) [inline src/Avalonia.Controls/Presenters/ItemVirtualizer.cs:170]: `scrollable.InvalidateScroll` was change to an event so we can't check to compare it to `null`, instead just check if its parent is a scroll content presenter (which is effectively what this check does).

File: Avalonia.sln (hunk 1 above)
Comment (Gillibald) [reply Avalonia.sln:207]: I thought I am the only one having this issue

File: src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs [hunk 4]
```diff
@@ -360,12 +362,17 @@ private void UpdateScrollableSubscription(IControl child)
                             .Subscribe(x => scrollable.CanVerticallyScroll = x),
                         this.GetObservable(OffsetProperty)
                             .Skip(1).Subscribe(x => scrollable.Offset = x),
-                        Disposable.Create(() => scrollable.InvalidateScroll = null));
+                        Disposable.Create(() => scrollable.ScrollInvalidated -= ScrollInvalidated));
```
Comment (MarchingCube) [inline src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs:365]: FYI: `Disposable.Create` has a variant that takes a parameter which allows you to easily drop allocations (lambda and closure) :)

Comment (grokys) [reply src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs:365]: TBH that whole block could probably be done without Rx but I was just refactoring the minimum to reduce noise in this PR...

File: src/Avalonia.Controls/ScrollViewer.cs [hunk 5]
```diff
@@ -10,6 +10,8 @@ namespace Avalonia.Controls
     /// </summary>
     public class ScrollViewer : ContentControl, IScrollable, IScrollAnchorProvider
     {
+        private static readonly Size s_defaultSmallChange = new Size(16, 16);
```
Comment (MarchingCube) [inline src/Avalonia.Controls/ScrollViewer.cs:13]: 16 again, should `ItemsPresenter` share this constant?

**Comment by grokys** [comment]
@danwalmsley actually I can probably port this to 0.9 without the breaking change if needed.

**Comment by grokys** [comment]
@MarchingCube feedback address, thank you!

**Review by MarchingCube — APPROVED** [review]
LGTM.

**Pull request merged**